drivers/l3gxxxx/l3gxxxx.c

```diff
@@ -0,0 +1,1163 @@
+
+    int res = L3GXXXX_OK;
+
+    uint8_t ig_cfg;
+    uint8_t ig_dur;
+    uint8_t ig_ths[6];
+    uint8_t ctrl3;
+    uint8_t ctrl5;
+
+    _acquire(dev);
+    res |= _read(dev, L3GXXXX_REG_IG_THS_XH, ig_ths, 6);
+    res |= _read(dev, L3GXXXX_REG_IG_CFG, &ig_cfg, 1);
+    res |= _read(dev, L3GXXXX_REG_IG_DURATION, &ig_dur, 1);
+    res |= _read(dev, L3GXXXX_REG_CTRL3, &ctrl3, 1);
+    res |= _read(dev, L3GXXXX_REG_CTRL5, &ctrl5, 1);
+    _release(dev);
+
+    cfg->x_low_enabled   = _GET_REG_BIT(ig_cfg, L3GXXXX_INT1_X_LOW);
+    cfg->x_high_enabled  = _GET_REG_BIT(ig_cfg, L3GXXXX_INT1_X_HIGH);
+
+    cfg->y_low_enabled   = _GET_REG_BIT(ig_cfg, L3GXXXX_INT1_Y_LOW);
+    cfg->y_high_enabled  = _GET_REG_BIT(ig_cfg, L3GXXXX_INT1_Y_HIGH);
+
+    cfg->z_low_enabled   = _GET_REG_BIT(ig_cfg, L3GXXXX_INT1_Z_LOW);
+    cfg->z_high_enabled  = _GET_REG_BIT(ig_cfg, L3GXXXX_INT1_Z_HIGH);
+
+    cfg->x_threshold     = msb_lsb_to_type(uint16_t, ig_ths, 0);
+    cfg->y_threshold     = msb_lsb_to_type(uint16_t, ig_ths, 2);
+    cfg->z_threshold     = msb_lsb_to_type(uint16_t, ig_ths, 4);
+
+    cfg->filter          = _GET_REG_VALUE(ctrl5, L3GXXXX_IG_SEL);
+
+    cfg->and_or          = _GET_REG_BIT(ig_cfg, L3GXXXX_INT1_AND_OR);
+    cfg->latch           = _GET_REG_BIT(ig_cfg, L3GXXXX_INT1_LATCH);
+
+    cfg->wait            = _GET_REG_BIT(ig_dur, L3GXXXX_INT1_WAIT);
+    cfg->duration        = _GET_REG_VALUE(ig_dur, L3GXXXX_INT1_DURATION);
+
+    cfg->counter_mode    = 0;
+
+    return res;
+}
+
+#endif /* IS_USED(MODULE_L3GXXXX_CONFIG) */
+
+int l3gxxxx_reg_read(const l3gxxxx_t *dev,
+                     uint8_t reg, uint8_t *data, uint8_t len)
+{
+    int res;
+
+    _acquire(dev);
+    res = _read(dev, reg, data, len);
+    _release(dev);
+
+    return res;
+}
+
+int l3gxxxx_reg_write(const l3gxxxx_t *dev,
+                      uint8_t reg, const uint8_t *data, uint8_t len)
+{
+   int res;
+
+    _acquire(dev);
+    res = _write(dev, reg, data, len);
+    _release(dev);
+
+    return res;
+}
+
+int l3gxxxx_reg_update(const l3gxxxx_t *dev,
+                       uint8_t reg, uint8_t mask, uint8_t val)
+{
+   int res;
+
+    _acquire(dev);
+    res = _update(dev, reg, mask, val);
+    _release(dev);
+
+    return res;
+}
+
+/** Functions for internal use only */
+
+#define L3GXXXX_I2C_L3GD20X_ADDR_1  (0x6a)      /* L3GD20x I2C address 1 */
+#define L3GXXXX_I2C_L3GD20X_ADDR_2  (0x6b)      /* L3GD20x I2C address 2 */
+#define L3GXXXX_I2C_X3G42XXD_ADDR_1 (0x68)      /* x3G42xxD I2C address 1 */
+#define L3GXXXX_I2C_X3G42XXD_ADDR_2 (0x69)      /* x3G42xxD I2C address 2 */
+
+/**
+ * @brief   Check the chip ID to test whether sensor is available
+ * @note    Communication bus has to be acquired when enter.
+ */
+static int _is_available(l3gxxxx_t *dev)
+{
+    DEBUG_DEV("", dev);
+
+    int res;
+
+    uint8_t chip_id;
+
+    /* try to read the chip id from L3GXXXX_REG_WHO_AM_I */
+    res = _read(dev, L3GXXXX_REG_WHO_AM_I, &chip_id, 1);
```

Comment:
Is there some power-up delay? On `stm32f429i-disco` this would always read `0xFF` (or 0) when `ENABLE_DEBUG` is *not* set, but work when set, so looks like the initial debug print is delay enough.

What works is

```C
    uint8_t tries = 255;

    /* try to read the chip id from L3GXXXX_REG_WHO_AM_I */
    do {
        res = _read(dev, L3GXXXX_REG_WHO_AM_I, &chip_id, 1);
    } while ((chip_id == 0xFF || chip_id == 0x0) && --tries);
```

Reply:
Hm, there is some power-up delay, but only for the gyros that are switched off by default after power-up because the sensor is in power-down mode. But the communication interfaces are of course always active to be able to communicate with the sensor also in power-down mode. There is no power-up delay is specified for the digital block in the datasheet.

In the initial version of the driver I had a delay of 100 ms at initialization, but this caused the dependency on `xtimer`. Now that we have `ztimer_msec`, we could add such a delay again, or we could try to read a valid ID several times before aborting.

Reply:
> or we could try to read a valid ID several times before aborting.

Ah ok, I see. I have overlooked your suggestion with the retries. So I guess that you prefer this approach. But the `ztimer_msec` could also be usedul in `l3gxxxx_power_up` to wait for the gyros. At the moment, switching from power-down to normal mode results in some invalid measurements until the gyros are stable.

Reply:
It's not wrong to use a timer, I just thought

```patch
diff --git a/drivers/l3gxxxx/l3gxxxx.c b/drivers/l3gxxxx/l3gxxxx.c
index bc5240d2d3..6caed8d47c 100644
--- a/drivers/l3gxxxx/l3gxxxx.c
+++ b/drivers/l3gxxxx/l3gxxxx.c
@@ -880,10 +880,12 @@ static int _is_available(l3gxxxx_t *dev)
 
     int res;
 
-    uint8_t chip_id;
+    uint8_t chip_id, tries = 255;
 
     /* try to read the chip id from L3GXXXX_REG_WHO_AM_I */
-    res = _read(dev, L3GXXXX_REG_WHO_AM_I, &chip_id, 1);
+    do {
+        res = _read(dev, L3GXXXX_REG_WHO_AM_I, &chip_id, 1);
+    } while ((chip_id == 0xFF || chip_id == 0x0) && --tries);
```

would be the most easy and robust solution.